EasyLM/models/llama/llama_train_ppo.py

```diff
@@ -0,0 +1,605 @@
+        'ppo/loss/policy': detach(pg_loss),
+        'ppo/loss/value': detach(vf_loss),
+        'ppo/loss/total': detach(loss),
+        'ppo/policy/ratios_mean': detach(masked_mean(ratio, cont_attn_mask)),
+        'ppo/policy/advantages_mean': detach(masked_mean(advantages, cont_attn_mask)),
+        'ppo/returns/mean': detach(masked_mean(returns, cont_attn_mask)),
+        'ppo/val/vpred': detach(masked_mean(new_cont_values, cont_attn_mask)),
+        'ppo/val/error': detach(masked_mean(jnp.square(new_cont_values - returns), cont_attn_mask)),
+        'ppo/val/mean': detach(masked_mean(old_cont_values, cont_attn_mask)),
+    }
+    return loss, stats
+
+def compute_advantages(values, rewards, mask):
+    lastgaelam = 0
+    advantages_reversed = []
+    gen_len = mask.shape[1]
+    values = values * mask
+    rewards = rewards * mask
+    if FLAGS.whiten_rewards:
+        rewards = whiten(rewards, mask, shift_mean=False)
+    for t in reversed(range(gen_len)):
+        nextvalues = values[:, t + 1] if t < gen_len - 1 else jnp.zeros_like(values[:, t])
+        delta = rewards[:, t] + FLAGS.gamma * nextvalues - values[:, t]
+        lastgaelam = delta + FLAGS.gamma * FLAGS.lam * lastgaelam
+        advantages_reversed.append(lastgaelam)
+    advantages = jnp.stack(advantages_reversed[::-1], axis=1)
+    returns = advantages + values
+    advantages = whiten(advantages, mask, shift_mean=True)
+    return advantages, returns
+
+def ppo_step(
+    policy_train_state, reference_train_state, value_train_state, reward_train_state,
+    policy_model, reference_model, value_model, reward_model,
+    rng, batch,
+):
+    rng_generator = JaxRNG(rng)
+    batch = with_sharding_constraint(batch, PS(('dp', 'fsdp')))
+
+    prompt_input_ids, prompt_attn_mask = batch['prompt_input_ids'], batch['prompt_attn_mask']
+    reward_prompt_input_ids, reward_prompt_attn_mask = batch['reward_prompt_input_ids'], batch['reward_prompt_attn_mask']
+    PL = prompt_input_ids.shape[1]
+
+    timing = dict()
+    t0 = time.time()
+
+    # rollout from current policy
+    t = time.time()
+    pad_token_id = 0
+    bos_token_id = 1
+    eos_token_id = 2
+    generation_config = GenerationConfig(
+        do_sample=True,
+        temperature=FLAGS.temperature,
+        pad_token_id=pad_token_id,
+        bos_token_id=bos_token_id,
+        eos_token_id=eos_token_id,
+        max_new_tokens=FLAGS.max_continuation_len,
+        # forced_eos_token_id=eos_token_id,
+    )
+    outputs = policy_model.generate(
+        input_ids=prompt_input_ids,
+        attention_mask=prompt_attn_mask,
+        generation_config=generation_config,
+        params=policy_train_state.params['params'],
+        prng_key=rng_generator(),
+    )
+    input_ids = outputs.sequences # (B, L)
+
+    # NOTE: This is a hack because generate() weirdly forces the last token to be 0 instead of 2
+    last_token_index = jnp.argmax(jnp.cumsum(jnp.where(input_ids == pad_token_id, 0, 1), axis=1), axis=1) # (B)
```

Comment:
Is this the same as `last_token_index = jnp.sum(jnp.where(input_ids == pad_token_id, 0, 1), axis=1)`?